Erictr/inkv2 (#195)
BonsaiAI/bonsaiai.github.io: source/includes/examples/_cartpole.html.md

## Inkling File

###### Types

```inkling2
type GameState {
    position: number,
    velocity: number,
    angle: number,
    rotation: number
}
```

The type `GameState` defines four fields - `position`, `velocity`, `angle`, and `rotation` — and assigns a type to each. This information is input from the simulation.

```inkling2
type Action {
    command: number<Left = 0, Right = 1>

> }

Reviewer: Confirming my understanding more than asking for an edit: this doesn't require braces, right?

type CartPoleConfig {
    episode_length: Number.Int8,
    deque_size: Number.UInt8
}
```

 The type `CartPoleConfig` defines two fields — `episode_length` and
 `deque_size` — and assigns each of them a type.  `episode_length` is a signed `Int8` because -1 is used for "run until pole drops".

###### Concept Graph

```inkling2
graph (input: GameState): Action {

    concept balance(input): Action {
        curriculum {
            source cartpole_simulator

            lesson balancing {
                constraint {
                    episode_length: -1,
                    deque_size: 1
                }
            }
        }
    }
    
    output balance
}
```

The concept is named `balance`, and it takes input from the simulator. That input is of type `GameState`. The balance concept outputs the move the AI should make in the simulator. This output is of type `Action`.

The curriculum defines one lesson, called `balancing`. It constrains the simulator's configuration such that the `episode_length` is -1 and the `deque_size` is 1 for all episodes.

###### Simulator

```inkling2
simulator cartpole_simulator(action: Action, config: CartPoleConfig): GameState {
}
```

> Simulator `cartpole_simulator` receives inputs of two types. The first, `Action`, specifies the action that the AI will take in the simulation. The second, `CartPoleConfig`, specifies the configuration of the simulation. The simulator outputs a value of type `GameState` which contains the state of the simulator at the end of an episode.

Reviewer: ```suggestion
Simulator `cartpole_simulator` receives inputs of two types. The first, `Action`, specifies the action that the AI will take in the simulation. The second, `CartPoleConfig`, specifies the configuration of the simulation. The simulator outputs a value of type `GameState` which contains the state of the simulator at each iteration.
```
